src/backend/distributed/master/master_stage_protocol.c

```diff
@@ -261,67 +258,41 @@ master_append_table_to_shard(PG_FUNCTION_ARGS)
 						errhint("Try running master_create_empty_shard() first")));
 	}
 
+	BeginOrContinueCoordinatedTransaction();
+
 	/* issue command to append table to each shard placement */
 	foreach(shardPlacementCell, shardPlacementList)
 	{
 		ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(shardPlacementCell);
-		char *workerName = shardPlacement->nodeName;
-		uint32 workerPort = shardPlacement->nodePort;
-		List *queryResultList = NIL;
+		MultiConnection *connection = GetPlacementConnection(FOR_DML, shardPlacement,
+															 NULL);
+		PGresult *queryResult = NULL;
+		int executeCommand = 0;
 
 		StringInfo workerAppendQuery = makeStringInfo();
 		appendStringInfo(workerAppendQuery, WORKER_APPEND_TABLE_TO_SHARD,
 						 quote_literal_cstr(shardQualifiedName),
 						 quote_literal_cstr(sourceTableName),
 						 quote_literal_cstr(sourceNodeName), sourceNodePort);
 
-		/* inserting data should be performed by the current user */
-		queryResultList = ExecuteRemoteQuery(workerName, workerPort, NULL,
-											 workerAppendQuery);
-		if (queryResultList != NIL)
-		{
-			succeededPlacementList = lappend(succeededPlacementList, shardPlacement);
-		}
-		else
+		executeCommand = ExecuteOptionalRemoteCommand(connection, workerAppendQuery->data,
+													  &queryResult);
+		PQclear(queryResult);
+		ForgetResults(connection);
+
+		if (executeCommand != 0)
 		{
-			failedPlacementList = lappend(failedPlacementList, shardPlacement);
+			MarkRemoteTransactionFailed(connection, false);
 		}
 	}
 
-	/* before updating metadata, check that we appended to at least one shard */
-	if (succeededPlacementList == NIL)
-	{
-		ereport(ERROR, (errmsg("could not append table to any shard placement")));
-	}
-
-	/* make sure we don't process cancel signals */
-	HOLD_INTERRUPTS();
-
-	/* mark shard placements that we couldn't append to as inactive */
-	foreach(failedPlacementCell, failedPlacementList)
-	{
-		ShardPlacement *placement = (ShardPlacement *) lfirst(failedPlacementCell);
-		uint64 placementId = placement->placementId;
-		char *workerName = placement->nodeName;
-		uint32 workerPort = placement->nodePort;
-
-		UpdateShardPlacementState(placementId, FILE_INACTIVE);
-
-		ereport(WARNING, (errmsg("could not append table to shard \"%s\" on node "
-								 "\"%s:%u\"", shardQualifiedName, workerName,
-								 workerPort),
-						  errdetail("Marking this shard placement as inactive")));
-	}
-
 	/* update shard statistics and get new shard size */
 	newShardSize = UpdateShardStatistics(shardId);
```

Comment:
Hm, we probably want to error out before we do more work like getting statistics. Maybe count successes.